### PR TITLE
CSTAR-630: Support case where client and server users are not the same

### PR DESCRIPTION
**Pull request by csplinter**

### CODE REVIEW

**Comment by csplinter** [comment]
Sometimes the cstar_perf tool host is not using the same linux user as the server under test host. This causes the DSE bootstrap process to fail because it tries to create the cstar_perf tool host home on the server under test host. 
```
[automaton@10.200.175.89] out: [10.200.241.2] run: rm -rf /home/automaton/fab/dse_builds/dse-5.0.0
[automaton@10.200.175.89] out: [10.200.241.2] run: mkdir -p /home/automaton/fab/dse_builds
[automaton@10.200.175.89] out: [10.200.241.2] out: mkdir: cannot create directory ‘/home/automaton’: Permission denied
[automaton@10.200.175.89] out: [10.200.241.2] out: 
[automaton@10.200.175.89] out: 
[automaton@10.200.175.89] out: 
[automaton@10.200.175.89] out: Fatal error: run() received nonzero return code 1 while executing!
[automaton@10.200.175.89] out: 
[automaton@10.200.175.89] out: Requested: mkdir -p /home/automaton/fab/dse_builds
[automaton@10.200.175.89] out: Executed: /bin/bash -l -c "mkdir -p /home/automaton/fab/dse_builds"
```

In order to solve this, we only use `os.path.expanduser` to get the `dse_build_local` location. This PR uses `os.path.join('~','fab','dse')` as the `dse_build_remote` location. This allows the path to properly expand when the given commands are run on the server under test host.

Tested: 
* cstar_docker setup 
* jenkins@blade server under test host + automaton@openstack cstar_perf tool host

**Comment by nastra** [comment]
There are certain cases where `os.path.join('~','fab','dse')` would't work. One case that I can think of is if such a path is passed to the JVM as a parameter.

**Comment by csplinter** [comment]
I think using $HOME could work. I will try it out

**Comment by csplinter** [comment]
I chose to add a local option to our get_dse_path and get_cassandra_path

fab_cassandra.py
```
def get_cassandra_path(local=True):
    return os.path.expanduser('~/fab/cassandra') if local else os.path.join(fab.run('pwd'), 'fab', 'cassandra')


def get_bin_path(local=True):
    return os.path.join(get_cassandra_path(local=local), 'bin')
```

fab_dse.py
```
def get_dse_path(local=True):
    return os.path.expanduser("~/fab/dse") if local else os.path.join(fab.run('pwd'), 'fab', 'dse')


def get_dse_conf_path(local=True):
    return os.path.join(get_dse_path(local=local), 'resources', 'dse', 'conf')


def get_cassandra_path(local=True):
    return os.path.join(get_dse_path(local=local), 'resources', 'cassandra')
```

File: tool/cstar_perf/tool/fab_cassandra.py [hunk 1]
```diff
@@ -44,11 +44,14 @@
 # Additional git remotes can be specified in this file
 GIT_REMOTES_FILE = os.path.join(os.path.expanduser("~"), ".cstar_perf", "git_remotes.json")
 
-def get_cassandra_path():
-    return os.path.expanduser('~/fab/cassandra')
 
-def get_bin_path():
-    return os.path.expanduser('~/fab/cassandra/bin')
+def get_cassandra_path(local=True):
+    return os.path.expanduser('~/fab/cassandra') if local else os.path.join(fab.run('pwd'), 'fab', 'cassandra')
```
Comment (nastra) [inline tool/cstar_perf/tool/fab_cassandra.py]: so can we always assume that `fab.run('pwd')` will give us the user's home directory? What if code is called while the user is outside of his home directory? Will stuff still work? I also think that **$HOME** should always work

Comment (csplint) [inline tool/cstar_perf/tool/fab_cassandra.py]: Using **$HOME** failed for all fab.get and fab.put commands, though the path does look strange
```
download: <file obj> <- /home/cstar/$HOME/fab/dse/resources/cassandra/conf/cassandra.yaml
```
I can look into this again in my AM

Comment (csplinter) [inline tool/cstar_perf/tool/fab_cassandra.py]: Using **$HOME** failed for all fab.get and fab.put commands, though the path does look strange
```
download: <file obj> <- /home/cstar/$HOME/fab/dse/resources/cassandra/conf/cassandra.yaml
```
I can look into this again in my AM

Comment (nastra) [inline tool/cstar_perf/tool/fab_cassandra.py]: how did the code look like that you were using?

Comment (csplinter) [inline tool/cstar_perf/tool/fab_cassandra.py]: I just had 
```
def get_dse_path():
    return os.path.join('$HOME', 'fab', 'dse')
```
Tried to retest today but http://repository.apache.org/ is timing out so I can't build. Will try again in a few hours

**Comment by csplinter** [comment]
The behavior here with the fabric.operations.get function
```
download: <file obj> <- /home/cstar/$HOME/fab/dse/resources/cassandra/conf/cassandra.yaml
```
Is caused by the fact that they are joining `home = ftp.normalize('.')` with the `remote_path`. In our case, the home is `/home/cstar` and the remote path is `$HOME/fab/dse/resources/cassandra/conf/cassandra.yaml`. See the [fabric source] (https://github.com/fabric/fabric/blob/master/fabric/operations.py/#L554), I am still investigating workarounds

**Comment by nastra** [comment]
> Regarding the first solution of using `os.path.join('~','fab','dse')` and the case of using such a path as a JVM parameter, could we do something similar to what we do [here](https://github.com/datastax/cstar_perf/blob/master/tool/cstar_perf/tool/fab_dse.py/#L190-L191)?

The approach from https://github.com/datastax/cstar_perf/blob/master/tool/cstar_perf/tool/fab_dse.py/#L190-L191 sounds like a reasonable workaround

**Comment by csplinter** [comment]
@nastra this is ready for another review

- `get_dse_path` changed to use `os.path.join('$HOME','fab','dse')`
- fab.get and fab.put calls changed to use `.replace('$HOME','~')` if using `get_dse_path`. This includes calls to `product.get_cassandra_path` which calls `get_dse_path` when the product = dse
- See [this comment] (https://github.com/datastax/cstar_perf/pull/238#issuecomment-235693814) for explanation of why we have to `.replace('$HOME','~')` for the fab.get and fab.put calls
- I did not change `fab_cassandra.get_cassandra_path` as I do not anticipate the Tech Cert team will test open source Cassandra and this fix is specific to our scenario. If it is requested later, we can always add it to the Cassandra code as well.
- I tested w/cstar_docker and w/client + server that have different users

**Comment by nastra** [comment]
+1

**Pull request merged**